### DISCUSSION TITLE
2D Pixel perfect game with smooth camera movement

### DISCUSSION THREAD

**Discussion by gcardozo123** [MonoGame/MonoGame#8617] [Graphics]
Hello 👋 
I'm trying to make a pixel perfect platformer with smooth camera movement. I'm using Monogame + Nez, and once I attached a `FollowCamera` to the player, its movement became quite jiterry. I noticed the jitter changes depending on the framerate: if I lock it to 30fps the player character jitter stops, but then the tilemap slopes look super jittery. And I already tried to call `Transform.RoundPosition()` during `Update` in all critical things (player entity, player SpriteAnimator, camera, etc). 

What am I missing? 😔

![jitter](https://github.com/user-attachments/assets/a1b69815-afb6-4d79-b499-0c0e5bdf76b2)

<img width="1192" alt="image" src="https://github.com/user-attachments/assets/64e20bb7-6f50-4b08-8bb3-814d0e6c891d" />

[Edit]

Probably relevant to mention: I'm using Nez `Scene.SceneResolutionPolicy.ShowAllPixelPerfect` which limits scaling to integer values (to ensure rendering is pixel perfect)

**Comment by gcardozo123** [comment]
Alright, I found a workaround: I can setup fixed timestep and use a camera movement lerp that doesn't cause the issue. This is okay for now, but it would be nice to find the root cause, it's likely the camera movement updating in subpixels not getting along with the character animation and position updates. 🤔

**Comment by Apostolique** [comment]
It looks like your camera gets updated before the character moves so you're looking at the position that the character was last frame.

**Reply by gcardozo123** [reply]
I don't really get why that would cause jittery movement, but you're absolutely right. This was the issue:
```csharp
public class Player : Entity
{
//...lots of other stuff in here...

    public override void Update()
    {
          base.Update(); // Update Player components
          //Handle Player movement...
    }
}
```
that `base.Update();` call would cause the `FollowCamera` to move before the player, so moving that call to the end of the function resolved this issue. Thanks a lot! 🎉

**Comment by ThomasFOG** [comment]
You might want to ask that question on the Nez discord, it seems more related to using Nez components than it is to using to MonoGame.

It's likely a matter of who's updated first between your camera and your character's position. It looks like to me that the camera isn't last updated.

**Reply by gcardozo123** [reply]
Yes, I should probably have posted on Nez, my bad. And yes the issue was the camera was using an outdated player position as I mentioned in the reply to Apostolique above. Thanks anyway :)

**Comment by gcardozo123** [comment]
Resolved by @Apostolique comment

**Comment by Vlada3333** [comment]
thanks for the information!
